### PR TITLE
Add count parameter to history endpoints

### PR DESCRIPTION
**Pull request by noddi**
PR for feature from #1266 (first comment).

### CODE REVIEW

File: indexing-service/src/main/java/io/druid/indexing/overlord/http/OverlordResource.java [hunk 1]
```diff
@@ -219,10 +222,21 @@ public Response setWorkerConfig(
   @Path("/worker/history")
   @Produces(MediaType.APPLICATION_JSON)
   public Response getWorkerConfigHistory(
-      @QueryParam("interval") final String interval
+      @QueryParam("interval") final String interval,
+      @QueryParam("count") final Integer count
   )
   {
     Interval theInterval = interval == null ? null : new Interval(interval);
+    if (theInterval == null && count != null && count > 0) {
```
Comment (drcrallen) [inline indexing-service/src/main/java/io/druid/indexing/overlord/http/OverlordResource.java]: suggest returning bad request if count is less then or equal to zero.

File: server/src/main/java/io/druid/server/audit/SQLAuditManager.java [hunk 2]
```diff
@@ -207,4 +217,82 @@ public AuditEntry map(int index, ResultSet r, StatementContext ctx)
     );
   }
 
+  @Override
+  public List<AuditEntry> fetchAuditHistory(final String key, final String type, int limit) {
```
Comment (drcrallen) [inline server/src/main/java/io/druid/server/audit/SQLAuditManager.java]: can these two functions be refactored a bit to not duplicate so much code?

File: server/src/main/java/io/druid/server/http/CoordinatorDynamicConfigsResource.java [hunk 3]
```diff
@@ -90,10 +91,21 @@ public Response setDynamicConfigs(final CoordinatorDynamicConfig dynamicConfig,
   @Path("/history")
   @Produces(MediaType.APPLICATION_JSON)
   public Response getDatasourceRuleHistory(
-      @QueryParam("interval") final String interval
+      @QueryParam("interval") final String interval,
+      @QueryParam("count") final Integer count
   )
   {
     Interval theInterval = interval == null ? null : new Interval(interval);
+    if (theInterval == null && count != null && count > 0) {
```
Comment (drcrallen) [inline server/src/main/java/io/druid/server/http/CoordinatorDynamicConfigsResource.java]: suggest bad request if count less than or equal to zero

File: server/src/main/java/io/druid/server/http/RulesResource.java [hunk 4]
```diff
@@ -108,10 +108,15 @@ public Response setDatasourceRules(
   @Produces(MediaType.APPLICATION_JSON)
   public Response getDatasourceRuleHistory(
       @PathParam("dataSourceName") final String dataSourceName,
-      @QueryParam("interval") final String interval
+      @QueryParam("interval") final String interval,
+      @QueryParam("count") final Integer count
   )
   {
     Interval theInterval = interval == null ? null : new Interval(interval);
+    if (theInterval == null && count != null && count > 0) {
```
Comment (drcrallen) [inline server/src/main/java/io/druid/server/http/RulesResource.java]: same thing about count

File: server/src/test/java/io/druid/server/audit/SQLAuditManagerTest.java [hunk 5]
```diff
@@ -135,6 +134,90 @@ public void testFetchAuditHistory() throws IOException
     Assert.assertEquals(entry, auditEntries.get(1));
   }
 
+  @Test
```
Comment (drcrallen) [inline server/src/test/java/io/druid/server/audit/SQLAuditManagerTest.java]: Can you add tests for the count <=0 case? The docs do not clarify what happens in such a case and a unit test would be helpful.

File: server/src/test/java/io/druid/server/http/RulesResourceTest.java [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+
+    Response response = rulesResource.getDatasourceRuleHistory("datasource1", interval, null);
+    List<AuditEntry> rulesHistory = (List) response.getEntity();
+    Assert.assertEquals(2, rulesHistory.size());
+    Assert.assertEquals(entry1, rulesHistory.get(0));
+    Assert.assertEquals(entry2, rulesHistory.get(1));
+
+    EasyMock.verify(auditManager);
+  }
+}
```
Comment (drcrallen) [inline server/src/test/java/io/druid/server/http/RulesResourceTest.java]: newline

**Comment by drcrallen** [comment]
@noddi This looks really good! a few comments and I think it'll be ready to go soon.

**Comment by noddi** [comment]
@drcrallen thanks for code review!

File: server/src/test/java/io/druid/server/audit/SQLAuditManagerTest.java (hunk 5 above)
Comment (drcrallen) [inline server/src/test/java/io/druid/server/audit/SQLAuditManagerTest.java]: Note this was really meant as a general comment and not really meaning "in this exact file"

**Comment by noddi** [comment]
Changes:
- *fetchAuditHistory* method now throws *IllegalArgumentException* if limit is <= 0
- refactored *fetchAuditHistory* method
- removed condition *count <= 0* in http methods, now *IllegalArgumentException* is caught and HTTP 400 returned in response
- added tests in *SQLAuditManagerTest*, *RulesResourceTest* for *count <= 0*

File: server/src/test/java/io/druid/server/http/RulesResourceTest.java (hunk 6 above)
Comment (noddi) [inline server/src/test/java/io/druid/server/http/RulesResourceTest.java]: Done.

File: server/src/main/java/io/druid/server/audit/SQLAuditManager.java [hunk 7]
```diff
@@ -207,4 +219,63 @@ public AuditEntry map(int index, ResultSet r, StatementContext ctx)
...
+      throws IllegalArgumentException
+  {
+    return fetchAuditHistoryLastEntries(null, type, limit);
+  }
+
+  private List<AuditEntry> fetchAuditHistoryLastEntries(final String key, final String type, int limit)
+      throws IllegalArgumentException
+  {
+    final int theLimit = getLimit(limit);
+    String queryString = "SELECT payload FROM %s WHERE type = :type";
```
Comment (drcrallen) [inline server/src/main/java/io/druid/server/audit/SQLAuditManager.java]: Can you do the String.format right here?

Comment (noddi) [inline server/src/main/java/io/druid/server/audit/SQLAuditManager.java]: Done.

**Comment by drcrallen** [comment]
Cool, please squash. I'm :+1: after squash and travis pass.

**Comment by fjy** [comment]
@noddi Thanks for all the awesome contributions!

**Comment by drcrallen** [comment]
Bouncing for travis

**Comment by noddi** [comment]
@fjy Thanks!

**Pull request merged**